DCP-3538: Improvement - Maintenance & refactor (#2)
Distributive-Network/use-dcp-worker: README.md

The hook accepts a single object with the following parameters:
- `identity?: Keystore`: A Keystore object (`dcp.wallet.Keystore`) which is passed to the Worker constructor and set as the Worker's identity when communicating over the network. If a Keystore is not provided, an arbitrary one will be generated.
- `useLocalStorage?: boolean = true`:  A flag to toggle the use of the browser's local storage. The `workerOptions` object is the entity to be saved to local storage and is updated accordingly when calling `setWorkerOptions`.

> - `workerOptions: object`: The contents of this object will override the default values coming from the worker configuration coming from `dcpConfig`(provided by `dcp-client`, worker node sourced from `dcp-worker`). The only required property of the `workerOptions` object needed to provide is a `paymentAddress`. The following properties describe the worker options object configuring the DCP Worker.

Reviewer: **suggestion:** Clarify the requirement of the `paymentAddress` option.

**nitpick:** Prefer using `:` rather than `.` when starting lists.

```suggestion
- `workerOptions: object`: The contents of this object will override the default values coming from the worker configuration coming from `dcpConfig`(provided by `dcp-client`, worker node sourced from `dcp-worker`). The only required property of `workerOptions` is `paymentAddress`. The following properties describe the worker options object configuring the DCP Worker:
```

    - `fetchData: []`: A list of origins that are allowed to fetch input set data from.
    - `sendResults: []`: A list of origins that are allowed to send results to.
  - `minimumWage?: object`: The minimum payout per slice the worker will accept from a job. Will default with the following structure:
    - `CPU: number = 0`
    - `GPU: number = 0`
    - `in: number = 0`
    - `out: number = 0`
  - `computeGroups?: []`: List of compute groups the worker is in and the authorization to join them. A compute group is to be described as `{ joinKey: 'exampleGroup', joinSecret: 'password' }`.
  - `leavePublicGroup?: boolean = false`: A flag that controls if the worker should omit fetching work from the public compute group. If not defined, this flag is evaluated to _false_.
  - `jobAddresses?: []`: If populated, worker will only fetch slices from jobs corresponding to the job addresses in this list.
  - `maxWorkingSandboxes?: integer | undefined`: Maximum number of sandboxes allowed to do work. If `undefined`, then the Supervisor will determine a safe limit, based off of machine hardware.

>   - `paymentAddress: Keystore | Address | String`: A Keystore, Address (`dcp.wallet.Address`) or string identifying a DCP Bank Account to deposit earned DCCs.

Reviewer: **suggestion:** I think the mandatory property should be listed first so that folks looking for the bare minimum can more easily parse it from skimming the text.

**suggestion:** I there's a difference between `String` and `string` given how `new String()` and `String()` work. Given that you're also using `boolean` rather than `Boolean`, we should opt for the more intuitive representation of the type.

```suggestion
  - `paymentAddress: Keystore | Address | string`: A Keystore, Address (`dcp.wallet.Address`) or string identifying a DCP Bank Account to deposit earned DCCs.
```

Author: Ive updated it so its: string | Address | Keystore. I believe string is preferable over String

Reviewer: https://github.com/Distributive-Network/use-dcp-worker/pull/2#discussion_r1293642571

> I'd also like to fully address the **first half** of https://github.com/Distributive-Network/use-dcp-worker/pull/2#discussion_r1245730062 as well..

My bad, I wasn't being clear. I was referring to this part of the suggestion:

> suggestion: I think the mandatory property should be listed first so that folks looking for the bare minimum can more easily parse it from skimming the text.

i.e., Order `paymentAddress` first in the docs, rather than near the end/mixed in with the other parameters.

**question:** Should I split my suggestions in to multiple comments to help avoid this confusion in the future?

Author: > Should I split my suggestions in to multiple comments to help avoid this confusion in the future?

No this is my fault I had misinterpreted what your comment was - Sorry! Ill make this change.

Note: Learn more about `Keystore` and `Address` in our [Wallet API documentation](https://docs.dcp.dev/specs/wallet-api.html).

## Returns
This hook returns an object with the following properties:
- `worker: Worker`: Refer to the [Worker API documentation](https://docs.dcp.dev/specs/worker-api.html).
- `workerState: object`: Stores status of worker states. Stored globally and preseved between component updates.
  - `isLoaded: boolean`: True once the worker is properly initialized.
  - `working: boolean`: True if the worker is doing work, false otherwise.
  - `willWork: boolean`: True when the worker is starting to do work, false when the worker is stopping.
  - `fetching: boolean`: True when the worker is fetching for slices to compute, false otherwise.
  - `submitting: boolean`: True when the worker is submitting results to the scheduler, false otherwise.
  - `error: Error | boolean`: Set when a worker error has occured, false otherwise.
  - `workingSandboxes: integer`: Number of sandboxes currently doing work.
- `workerStatistics: object`: Stores a global count of worker statistics for a browser session. Stored globally and preseved between component updates.

>   - `slices: number`: Number of slices completed.

Reviewer: **suggestion:** `integer` isn't a valid primitive type in Typescript, but I get why it's described this way.

In other places, we use `number` even though `integer` could also work. For consistency and what I've normally seen, `change all instances of `integer` to `number`.

```suggestion
  - `workingSandboxes: number`: Number of sandboxes currently doing work.
- `workerStatistics: object`: Stores a global count of worker statistics for a browser session. Stored globally and preseved between component updates.
  - `slices: number`: Number of slices completed.
```

- `remove_byKey(key):` Removes all origins for all purposes under the _key_. A `null` _key_ is not accepted, must be a string.
- `getAllowList(purpose, key):` returns a list of origins under the _purpose_ and _key_. A `null` _purpose_ is not accepted, must be a string. Previously added origins under `null` _purpose_ and/or _key_ will match any _purpose_ and/or _key_, respectively.
- `isAllowed(origin, purpose, key):` returns `true` if _origin_ is allowed under the _purpose_ and _key_. A `null` _purpose_ is not accepted, must be a string. Previously added origins under `null` _purpose_ and/or _key_ will match any _purpose_ and/or _key_, respectively.
# Change log
- __1.0.0__ - Feb 28, 2023
  - Inital Release
- __1.0.1__ - Mar 1, 2023
  - Added src/ directory to published files to support source map
- __1.0.2__ - Mar 23, 2023
  - local storage will only save paymentAddress and maxWorkingSandbox props
  - workerOptions source always coming from dcpConfig
  - added delay between quick worker fetching states
  - Quality of life + maintainability improvements

> - __1.1.0__ - May

Reviewer: **suggestion:** This update is a major semver change as they break existing apps if people simply update their version and don't do anything else.

e.g., They'll need to migrate their usage of the hook from v1 to v2.

```suggestion
- __2.0.0__ - May
```

The version in `package.json` should also be changed.

  - now returns the worker
  - proper handling of race-condition in constructing the worker when/if the hook is executed multiple times at once
  - workerOptions passed to worker constructor is a Proxy now
    - needed reflect changes to paymentAddress and maxWorkingSandbox props to local storage and trigger component update/re-render
  - removed workerState.isLoaded since now we are returning the worker
    - !isLoaded === !worker
  - removed workerOptionsState, start/stop/toggleWorker, sandboxes, originManager
    - these are all accessible via Worker API
  - removed applyWorkerOptions since editing workerOptions should be done by directly mutating worker.workerOptions
  - improved error handling

>   - quality of life improvements

Reviewer: **suggestion:** Order the Changelog in reverse chronological order.

Most folks want to see the latest changes described first [^1]. e.g., On a new release, they'll want to see the new changes listed first in the file, not at the end of it after scrolling down.

[^1]: https://keepachangelog.com/en/1.0.0/#how

The hook accepts a single object with the following parameters:
- `identity?: Keystore`: A Keystore object (`dcp.wallet.Keystore`) which is passed to the Worker constructor and set as the Worker's identity when communicating over the network. If a Keystore is not provided, an arbitrary one will be generated.
- `useLocalStorage?: boolean = true`:  A flag to toggle the use of the browser's local storage. The `workerOptions` object is the entity to be saved to local storage and is updated accordingly when calling `setWorkerOptions`.

> - `workerOptions: object`: The contents of this object will override the default values coming from the worker configuration provided by the `dcp-worker` library (is this correct?? dcp-client library? I know the web-config service is responsible for dcpConfig.worker). The only required property of the `workerOptions` object needed to provide is a `paymentAddress`. The following properties describe the worker options object configuring the DCP Worker.

Reviewer 2: yea it comes from `dcp-worker/etc/dcp-worker-config.js`

  - `trustComputeGroupOrigins?: boolean = true`: Trust the scheduler to tell client about allowed origins for jobs in a compute group.

>   - `allowOrigins?: object`: Allow list permitting network access beyond DCP messages to services. This list is used only in setting up the DCP Worker. After the worker is constructed/loaded, the `originManager` is responsible for managing origins (see Managing Origins). Empty by default of course.

Reviewer: **suggestion:** Remove wording/tone that could be interpreted as condescending. 

It could be not obvious to some folks. Implying that it's a trivial thing to assume isn't super welcoming.

```suggestion
  - `allowOrigins?: object`: Allow list permitting network access beyond DCP messages to services. This list is used only in setting up the DCP Worker. After the worker is constructed/loaded, the `originManager` is responsible for managing origins (see Managing Origins). It's empty by default.
```
